File: src/python/comportamentais/chain_of_responsibility.py
# -*- coding: utf-8 -*-

# Esse padrão é útil quando você quer que múltiplos objetos tenham chance de processar um evento
class GUIComponent:
    def __init__(self, parent=None):
        self.parent = parent

    def handle_click(self, x, y):
        if self.parent:
            return self.parent.handle_click(x, y)
        print("Clique não tratado.")


class Button(GUIComponent):
    def __init__(self, x, y, w, h, parent=None):
        super().__init__(parent)
        self.x, self.y, self.w, self.h = x, y, w, h

    def handle_click(self, x, y):
        if self.x <= x <= self.x + self.w and self.y <= y <= self.y + self.h:
            print("Botão clicado!")
        else:
            super().handle_click(x, y)


class Panel(GUIComponent):
    def handle_click(self, x, y):
        print("Painel recebeu o clique, repassando...")
        super().handle_click(x, y)


class Window(GUIComponent):
    def handle_click(self, x, y):
        print("Janela recebeu o clique, repassando...")
        super().handle_click(x, y)


# Criando a cadeia
janela = Window()
painel = Panel(parent=janela)
botao = Button(10, 10, 100, 30, parent=painel)

# Teste
botao.handle_click(15, 20)  # Dentro do botão
botao.handle_click(200, 200)  # Fora do botão

File: src/cpp/comportamentais/chain_of_responsibility/ChainOfResponsibility.cpp

#include "ChainOfResponsibility.hpp"

#include <iostream>
GUIComponent::GUIComponent(std::shared_ptr<GUIComponent> parent) : parent_(std::move(parent)) {}

void GUIComponent::handleClick(int x, int y) {
    if (parent_) {
        parent_->handleClick(x, y);
    }
}

Button::Button(int x, int y, int width, int height, std::shared_ptr<GUIComponent> parent)
    : GUIComponent(parent), x_(x), y_(y), width_(width), height_(height) {}

void Button::handleClick(int x, int y) {
    if (x >= x_ && x < x_ + width_ && y >= y_ && y < y_ + height_) {
        std::cout << "Clicou no botão" << std::endl;
    } else {
        GUIComponent::handleClick(x, y);
    }
}

Panel::Panel(std::shared_ptr<GUIComponent> parent) : GUIComponent(parent) {}

void Panel::handleClick(int x, int y) {
    std::cout << "Clicou no painel" << std::endl;
    GUIComponent::handleClick(x, y);
}

Window::Window(std::shared_ptr<GUIComponent> parent) : GUIComponent(parent) {}

void Window::handleClick(int x, int y) {
    std::cout << "Clicou na janela" << std::endl;
    GUIComponent::handleClick(x, y);
}
